### PR TITLE
Config Flow

### PR DESCRIPTION
**Pull request by AaronDavidSchneider**
First try to implement a config flow.
Not working yet!

Edit: This is going to be a larger pull request. Many things need to be rewritten for this to work.

### CODE REVIEW

**Comment by mammuth** [comment]
Related issue: #30

**Comment by AaronDavidSchneider** [comment]
@mammuth: If you find some time, it would be nice, if you could give some feedback to this pull request.

Overview of working and not working stuff:
- [x] configuration via config flow and `configuration.yaml`
- [x] direct testing for valid credentials
- [ ] device_list in config_flow?! Is this even possible?
- [ ] change strings.json, translate de.json, add function headers to every function
- [ ] more testing
- [ ] Documentation of setup in README.md
- [ ] automatic discovery of fritzbox and suggestion of config
- [ ] Fix `This integration has no devices.`: List all related entities

**Comment by mammuth** [comment]
> device_list in config_flow?! Is this even possible?

I saw form inputs of integers and strings in other integrations; I could imagine that lists are also possible.

I think we should keep this PR as small as possible and move tasks that can be done later into separate PRs. Do you agree?

This has two advantages: 
- This PR is easier to review
- We can integrate the basic config flow quicker (and can improve stuff like translations, auto discovery etc. later)

> automatic discovery of fritzbox and suggestion of config

This is actually a pretty cool idea!

**Comment by AaronDavidSchneider** [comment]
> > device_list in config_flow?! Is this even possible?
> 
> I saw form inputs of integers and strings in other integrations; I could imagine that lists are also possible.
> 
> I think we should keep this PR as small as possible and move tasks that can be done later into separate PRs. Do you agree?

I agree! 

> 
> This has two advantages:
> 
> * This PR is easier to review
> * We can integrate the basic config flow quicker (and can improve stuff like translations, auto discovery etc. later)

Later on, translations will be done by the home assistant team anyway.

> 
> > automatic discovery of fritzbox and suggestion of config
> 
> This is actually a pretty cool idea!

**Comment by AaronDavidSchneider** [comment]
> * [ ]  Fix `This integration has no devices.`: List all related entities

For future reference:
https://developers.home-assistant.io/docs/en/device_registry_index.html
will probably solve this issue.... We would therefore need to rename CONF_DEVICES back to local CONF_DEVICE_LIST... 
Maybe later...

File: custom_components/fritzbox_tools/const.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+"""Constants for the AdGuard Home integration."""
```
Comment (AaronDavidSchneider) [inline custom_components/fritzbox_tools/const.py]: Ups... copy pasted and forgot to change

File: custom_components/fritzbox_tools/sensor.py [hunk 2]
```diff
@@ -1,23 +1,26 @@
 """AVM Fritz!Box connectivitiy sensor"""
+import asyncio
 import logging
-from datetime import timedelta
 from collections import defaultdict
-import asyncio
+from datetime import timedelta
```
Comment (AaronDavidSchneider) [inline custom_components/fritzbox_tools/sensor.py]: automated ordering of inputs using isort. See https://developers.home-assistant.io/docs/en/development_guidelines.html

**Comment by AaronDavidSchneider** [comment]
Screenshot 2019-11-15 at 11 18 04

This is how the component is listed as Integration. I think my work on this pull request ist completed now. 
TODOs for future pull requests:
- [ ] automatic discovery
- [ ] moving tests from `async_setup_entry` to fritzbox_tools.is_ok(). This way the component only gets created if the configuration in either configuration.yaml or form is correct and the user gets more specific answers to what he did wrong.
- [ ] translations
- [ ] function headers and other cosmetics...
- [ ] automatic update of component when configuration.yaml is changed (at the moment you need to unload it first and then restart home assistant)

**Comment by mammuth** [comment]
The form is currently hard to understand without reading the instructions in the README I think. One way to improve it would be to introduce multiple steps (or do I misunderstand the concept of steps? Btw, why is the current step called _"user"_?).

This way, the first step could be the required fields (username, password + the two defaults for IP and port), the second step could be "Advanced configuration" with device profiles and port forwardings - or we could even use two separate steps for those features (might actually be smarter, because it allows us to have a separate description for each of those features; or is there a concept of sections or any way to add additional text to the form (I didn't see something like that in the docs)).

A question arises out of this: do we want to introduce multiple steps before releasing or do we want to keep it as is for now?

**Comment by AaronDavidSchneider** [comment]
Hey,
I really like your idea of multiple "steps"!

However the concept of steps in config_flow is something different. See: https://developers.home-assistant.io/docs/en/config_entries_config_flow_handler.html#defining-steps

I think this can be done some how. One needs to combine multiple show_form inside the user step.

I can try this and come back to you, if its too complicated.

**Comment by mammuth** [comment]
Ah, I understand it now. Well, then let's keep it as one large form for now (and maybe rephrase some of the placeholders so it's easier to understand).

Btw, I just noticed, that all updates are running on the `MainThread` on my test setup (which lead to a strange behavior after finishing the config flow: The UI was unresponsive for quite a while it seemed).

**Comment by AaronDavidSchneider** [comment]
Yes, I have noticed this as well. Is there a different way?
Feel free to change the form to what suits best!
I will try to find a way to have different forms

**Comment by AaronDavidSchneider** [comment]
> Maybe related: The integration page only lists device profile switches for me, although the other entities (sensor, port forward and guest wifi switch) are created and available 🤔
> image

Thats very strange! Does it occur everytime?
cant confirm this for my setup.

**Comment by AaronDavidSchneider** [comment]
If you have the time @mammuth: The plex integration uses multiple steps. This might help to implement a multistep form.

**Comment by mammuth** [comment]
> Thats very strange! Does it occur everytime? cant confirm this for my setup.

Yes, it only lists a couple of profile switches. I also completely deleted `.storage/` twice and freshly set up the integration. 

What just came to my mind: I've got device name clashes. Maybe that's related 🤔 But since the entities are created (just not attached to the integration's device) it's probably not the reason?

> The plex integration uses multiple steps. This might help to implement a multistep form.

Cool. Gonna check it out. But I think it's fine to do this later.

**Comment by mammuth** [comment]
@AaronDavidSchneider do you also have `devices: null` in `./storage/core.config_entries`? 
image

File: custom_components/fritzbox_tools/const.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+CONF_PROFILE_ON = 'profile_on'
+CONF_PROFILE_OFF = 'profile_off'
+CONF_HOMEASSISTANT_IP = 'homeassistant_ip'
+
+DEFAULT_PROFILE_OFF = 'Gesperrt'
+DEFAULT_HOST = '192.168.178.1'  # set to fritzbox default
+DEFAULT_PORT = 49000            # set to fritzbox default
+DEFAULT_USERNAME = ''           # set to fritzbox default?!
+DEFAULT_PROFILE_ON = None
```
[the remark this answers is not in the document]
Comment (AaronDavidSchneider) [reply custom_components/fritzbox_tools/const.py]: if we do so, the switches will be created even though the user might doesnt want them.

File: custom_components/fritzbox_tools/switch.py [hunk 4]
```diff
@@ -46,9 +47,9 @@
                     port_switches.append(
                         FritzBoxPortSwitch(fritzbox_tools, portmap, i, connection_type)
                     )
-        except:
+        except Exception as e:
             _LOGGER.error(
-                'Port switches could not be enabled. Check if your fritzbox is able to do port forwardings!',
+                f'Port switches could not be enabled. Check if your fritzbox is able to do port forwardings!\n {e}',
```
[the remark this answers is not in the document]
Comment (AaronDavidSchneider) [reply custom_components/fritzbox_tools/switch.py]: sounds good. Didnt know.

`except Exception:` instead of `except:`
is needed for flake8 or black (which is used by HA) (as far as I remember)

**Comment by AaronDavidSchneider** [comment]
> @AaronDavidSchneider do you also have `devices: null` in `./storage/core.config_entries`?
> image
> 

In `./storage/core.config_entries` you should have
```
"devices": [
                        "Helens-iPhone",
                        "Helens-iPad",
                        "Aarons-MacBook-Air",
                        "Aarons-iPhone",
                        "The-iPad",
                        "Helens-MacBook-Air",
                        "mac-n-42"
                    ],
```

> Hm, in `.storage/core.entity_registry` all entities do have the correct relation to the device (`"device_id": "459bf9fd44a1452bbbe9a1a6d9577e0c",`). 🤔

File: custom_components/fritzbox_tools/sensor.py [hunk 5]
```diff
@@ -38,6 +41,14 @@ def __init__(self, fritzbox_tools):
     def is_on(self) -> bool:
         return self._is_on
 
+    @property
+    def unique_id(self):
+        return self.fritzbox_tools.unique_id
```
[the remark this answers is not in the document]
Comment (AaronDavidSchneider) [reply custom_components/fritzbox_tools/sensor.py]: good idea. Should be changed accordingly.

**Comment by mammuth** [comment]
> I really don't understand why you get different results then I do.

Well, stupid me... I was running a testing HA a couple of versions old. The latest `dev` HA shows the properties fine... 🙈

File: custom_components/fritzbox_tools/const.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mammuth) [reply custom_components/fritzbox_tools/const.py]: Ah, I remember.

But shouldn't we rather add a boolean setting or something like this instead of mystically toggling the whole feature based on the name of `profile_on`?

In yaml I'd say we could do it like that: 
```yaml
fritzbox_tools:
   ...
   device_profiles:  # as soon as this is here, let's turn the feature on
      devices:  # optional
          - "Aaron's iPhone"
```
No clue whether/how this can be nicely done with config entries.

**Comment by mammuth** [comment]
Regarding the blocking of the `MainThread`: I just checked it out, the setup was previously also done in the main thread apparently. So nothing changed here. 

But the component seems to take way longer to set up with the changes in this branch (so that's way we notice it now). 🤔

File: custom_components/fritzbox_tools/const.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AaronDavidSchneider) [reply custom_components/fritzbox_tools/const.py]: I will try this tomorrow

[the remark this answers is not in the document]
Comment (AaronDavidSchneider) [reply custom_components/fritzbox_tools/const.py]: changed it @mammuth. Now you need to specify which devices you need.

**Comment by mammuth** [comment]
Cool. How do we go about the super slow setup? :thinking_face: Ignore and publish or should we try to debug it further?

**Comment by AaronDavidSchneider** [comment]
I don’t know... one can add jobs to the threadscheduler. I will try this tomorrow.
I don’t know if this will speed up the setup.

**Comment by AaronDavidSchneider** [comment]
I think I speeded up the setup. The idea is to outsource the adding of entries. This way the init of the component only tells the scheduler what has to be done.
My raspberry pi 2 with hassio doesn't suffer large delays.

**Comment by mammuth** [comment]
Cool, sounds good!
I'll try to test it again and release it within the next days.

**Comment by mammuth** [comment]
Awesome! I just released this as 1.5.0 on HACS. I'll test whether the update runs smoothly on my production setup now.
